File: JavaFiles/week1/PercolationStats.java
package week1;
import edu.princeton.cs.algs4.StdRandom;
import edu.princeton.cs.algs4.StdStats;

public class PercolationStats {
    private double[] trialsResults;

    public PercolationStats(int n, int trials) {
        if (n <= 0 || trials <= 0) {
            throw new IllegalArgumentException("n and trials must be greater than 0");
        }
        trialsResults = new double[trials];
        for (int i = 0; i < trials; i++) {
            Percolation perc = new Percolation(n);
            while (!perc.percolates()) {
                int row = StdRandom.uniform(1, n + 1);
                int col = StdRandom.uniform(1, n + 1);
                perc.open(row, col);
            }
            trialsResults[i] = (double) perc.numberOfOpenSites() / (n * n);
        }
    }

    public double mean() {
        return StdStats.mean(trialsResults);
    }

    public double stddev() {
        return StdStats.stddev(trialsResults);
    }

    public double confidenceLo() {
        return mean() - 1.96 * stddev() / Math.sqrt(trialsResults.length);
    }

    public double confidenceHi() {
        return mean() + 1.96 * stddev() / Math.sqrt(trialsResults.length);
    }

    
    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        int trials = Integer.parseInt(args[1]);
        PercolationStats stats = new PercolationStats(n, trials);

        System.out.println("Mean: " + stats.mean());
        System.out.println("Standard deviation: " + stats.stddev());
        System.out.println("95% confidence interval: [" + stats.confidenceLo() + ", " + stats.confidenceHi() + "]");
    }
}

File: week1/percolation_stats.py
import numpy as np
from percolation import Percolation

class PercolationStats:
    def __init__(self, n, trials):
        # Perform independent trials on an n-by-n grid
        self.trials_results = []
        for _ in range(trials):
            p = Percolation(n) 
            while not p.percolates():
                p.open(np.random.randint(1, p._n+1), np.random.randint(1, p._n+1))
            self.trials_results.append(p.count / (n * n))

    def mean(self):
        # Sample mean of percolation threshold
        return np.mean(self.trials_results)

    def stddev(self):
        # Sample standard deviation of percolation threshold
        return np.std(self.trials_results)

    def confidenceLo(self):
        # Low endpoint of 95% confidence interval
        return self.mean() - 1.96 * self.stddev() / np.sqrt(len(self.trials_results))

    def confidenceHi(self):
        # High endpoint of 95% confidence interval
        return self.mean() + 1.96 * self.stddev() / np.sqrt(len(self.trials_results))

# Test client
if __name__ == "__main__":

    stats = PercolationStats(200, 100)
    
    print("Mean: ", stats.mean())
    print("Standard deviation: ", stats.stddev())
    print("95% confidence interval: [", stats.confidenceLo(), ",", stats.confidenceHi(), "]")
